docs: external KMS configuration (#23600)
hashicorp/nomad: website/content/docs/configuration/keyring/azurekeyvault.mdx

---
layout: docs
page_title: Azure Key Vault - Keyring - Configuration
description: >-
  The Azure Key Vault keyring configures Nomad to use Azure Key Vault as the key material
  wrapping mechanism.
---

# `azurekeyvault` keyring

The AWS KMS keyring configures Nomad to use AWS KMS to wrap its keyring. This
example shows configuring Azure Key Vault seal through the Vault configuration

> file by providing all the required values:

Reviewer: ```suggestion
file by providing all the required values.
```

```hcl
seal "azurekeyvault" {
  active = true
  name   = "example"

  # fields specific to azurekeyvault
  tenant_id      = "46646709-b63e-4747-be42-516edeaf1e14"
  client_id      = "03dc33fc-16d9-4b77-8152-3ec568f8af6e"
  client_secret  = "DUJDS3..."
  vault_name     = "nomad"
  key_name       = "nomad_key"
}
```

## `azurekeyvault` parameters

These parameters apply to the `keyring` block in the Nomad configuration file:

- `tenant_id` `(string: <required>)`: The tenant id for the Azure Active

>   Directory organization. May also be specified by the `AZURE_TENANT_ID`
>   environment variable.

Reviewer: ```suggestion
  Directory organization. Alternately specify via the `AZURE_TENANT_ID` environment variable.
```

- `client_id` `(string: <required or MSI>)`: The client id for credentials to

>   query the Azure APIs.  May also be specified by the `AZURE_CLIENT_ID`
>   environment variable.

Reviewer: ```suggestion
  query the Azure APIs.  Alternately specify via the `AZURE_CLIENT_ID` environment variable.
```

- `client_secret` `(string: <required or MSI>)`: The client secret for

>   credentials to query the Azure APIs.  May also be specified by the
>   `AZURE_CLIENT_SECRET` environment variable.

Reviewer: ```suggestion
  credentials to query the Azure APIs.  Alternately specify via the `AZURE_CLIENT_SECRET` environment variable.
```

- `environment` `(string: "AZUREPUBLICCLOUD")`: The Azure Cloud environment API

>   endpoints to use. May also be specified by the `AZURE_ENVIRONMENT` environment

Reviewer: ```suggestion
  endpoints to use. Alternately specify via the `AZURE_ENVIRONMENT` environment
```

  variable.

- `vault_name` `(string: <required>)`: The Key Vault vault to use the encryption
  keys for encryption and decryption.

- `key_name` `(string: <required>)`: The Key Vault key to use for encryption and
  decryption.

- `resource` `(string: "vault.azure.net")`: The AZ KeyVault resource's DNS

>   Suffix to connect to.  May also be specified in the `AZURE_AD_RESOURCE`

Reviewer: ```suggestion
  Suffix to connect to.  Alternately specify via the `AZURE_AD_RESOURCE`
```

  environment variable.  Needs to be changed to connect to Azure's Managed HSM
  KeyVault instance type.

## Authentication

> Authentication-related values must be provided, either as environment variables
> or as configuration parameters.

Reviewer: ```suggestion
You must provide authentication-related values either as environment variables
or as configuration parameters.
```

Azure authentication values:

- `AZURE_TENANT_ID`
- `AZURE_CLIENT_ID`
- `AZURE_CLIENT_SECRET`
- `AZURE_ENVIRONMENT`
- `AZURE_AD_RESOURCE`

> ~> **Note:** If Nomad is hosted on Azure, Vault can use Managed Service

Author: Oops, caught me copy-and-pasting from Vault's docs.

> ~> **Note:** If Nomad is hosted on Azure, Vault can use Managed Service
> Identities (MSI) to access Azure instead of an environment and shared client id
> and secret. MSI must be [enabled][msi_enabled] on the VMs hosting Nomad, and it
> is the preferred configuration since MSI prevents your Azure credentials from
> being stored as clear text.

Reviewer: ```suggestion
~> **Note:** If Nomad is hosted on Azure, Vault can use Managed Service Identities (MSI) to access Azure instead of an environment and shared client id and secret. MSI must be [enabled][msi_enabled] on the VMs hosting Nomad. We recommend this configuration since MSI prevents your Azure credentials from being stored as clear text.
```

style guide: active voice (it is the preferred configuration -> We recommend this configuration)

> -> **Note:** If you are using a Managed HSM KeyVault, `AZURE_AD_RESOURCE` or the
> `resource` configuration parameter must be specified; usually this should point
> to `managedhsm.azure.net`, but could point to other suffixes depending on Azure
> environment.

Reviewer: ```suggestion
-> **Note:** If you are using a Managed HSM KeyVault, you must specify `AZURE_AD_RESOURCE` or the `resource` configuration parameter; usually this should point to `managedhsm.azure.net` but could point to other suffixes depending on Azure environment.
```
style guide: active voice

## Key rotation

This seal supports rotating keys defined in Azure Key Vault. Key metadata is
stored with the encrypted data to ensure the correct key is used during

> decryption operations. Set up Azure Key Vault with [key rotation][] using Azure
> Automation Account and Vault will recognize newly rotated keys.

Reviewer: ```suggestion
decryption operations. Set up Azure Key Vault with [key rotation][] using Azure
Automation Account so Vault recognizes newly rotated keys.
```
style guide: active voice
